nslsii/EPS_devices/valves_and_shutters.py

```diff
@@ -0,0 +1,333 @@
+
+        self._set_st = None
+        self.read_attrs = ['status']
+        self._state1_val = state1_val
+        self._state2_val = state2_val
+        self._state1_str = state1_str
+        self._state2_str = state2_str
+        self._num_retries = num_retries
+        self._retry_sleep_time = retry_sleep_time
+        if stop_str == state1_str:
+            self._stop_str = self._state1_str
+            self._resume_str = self._state2_str
+        elif stop_str == state2_str or stop_str is None:
+            self._stop_str = self._state2_str
+            self._resume_str = self._state1_str
+        else:
+            raise nslsiiValves_and_ShuttersValueError(
+                'the kwarg ``stop_str`` in the EpicsTwoStateDevice class'
+                '``__init__`` method needs to be None or needs to match one of'
+                'the kwargs ``state1_str`` or ``state2_str``')
+
+        self._cmd_map = {self._state1_str: self._state1_cmd,
+                         self._state2_str: self._state2_cmd}
+        self._target_map = {self._state1_str: self._state1_val,
+                            self._state2_str: self._state2_val}
+
+        self._time_fmtstr = '%Y-%m-%d %H:%M:%S'
+
+    _state1_cmd = FmtCpt(EpicsSignal,
+                         '{self.prefix}Cmd:{self._state1_pv_uid}-Cmd',
+                         string=True, kind='omitted')
+    _state2_cmd = FmtCpt(EpicsSignal,
+                         '{self.prefix}Cmd:{self._state2_pv_uid}-Cmd',
+                         string=True, kind='omitted')
+
+    status = Cpt(EpicsSignalRO, 'Pos-Sts', string=True)
+
+    enabled_status = Cpt(EpicsSignalRO, 'Enbl-Sts', string=True, kind='config')
+
+    _fail_to_state1 = FmtCpt(EpicsSignalRO,
+                             '{self.prefix}Sts:Fail{self._state1_pv_uid}-Sts',
+                             string=True, kind='omitted')
+    _fail_to_state2 = FmtCpt(EpicsSignalRO,
+                             '{self.prefix}Sts:Fail{self._state2_pv_uid}-Sts',
+                             string=True, kind='omitted')
+
+    def set(self, val):
+        if self._set_st is not None:
+            raise RuntimeError(
+                f'trying to set {self.name} while a set is in progress')
+
+        cmd_sig = self._cmd_map[val]
+        target_val = self._target_map[val]
+
+        st = self._set_st = DeviceStatus(self)
+        enums = self.status.enum_strs
+
+        def state_cb(value, timestamp, **kwargs):
+            value = enums[int(value)]
+            if value == target_val:
+                self._set_st._finished()
+                self._set_st = None
+                self.status.clear_sub(state_cb)
+
+        cmd_enums = cmd_sig.enum_strs
+        count = 0
+
+        def cmd_retry_cb(value, timestamp, **kwargs):
+            nonlocal count
+            value = cmd_enums[int(value)]
+            count += 1
+            if count > self._num_retries:
+                cmd_sig.clear_sub(cmd_retry_cb)
+                st._finished(success=False)
```

Comment:
I think at this point `st` and `self._set_st` are different: the former is a status object, and the latter is `None`.

It can be demonstrated by this snippet:
```py
In [1]: from ophyd.status import Status

In [2]: st1 = st2 = Status()

In [3]: st1
Out[3]: Status(obj=None, done=False, success=False)

In [4]: st2
Out[4]: Status(obj=None, done=False, success=False)

In [5]: st1 = None

In [6]: st1

In [7]: st2
Out[7]: Status(obj=None, done=False, success=False)
```

Reply:
thanks for the snippet @mrakitin this was useful for me to see the difference. I will leave off making any changes so that I can see @tacaswell's detailed discussion on race conditions

Reply:
I have resolved this, see response to @tacaswell below